### ISSUE TITLE
Numpy.array - wrong dtype for 'string enum'

### ISSUE THREAD

**Issue by alonme** [numpy/numpy#15767]
When creating a `numpy.array` of a class that inherits from both str and Enum the dtype is `dtype='<UX`
where X is the length of the string. the data saved is X chars from the string representation of the enum.

I would expect the minimal value to be object, so the Enum is preserved.

### Reproducing code example:

```python
>>> import numpy as np
>>> from enum import Enum
>>> class StrEnum(str, Enum):
...    A = 'a'
...    B = 'bb'
...
>>> np.array([StrEnum.A])
array(['S'], dtype='<U1')
>>> np.array([StrEnum.B])
array(['St'], dtype='<U2')
```

### Numpy/Python version information:
1.18.0 3.7.5 (default, Nov  1 2019, 02:16:23)
[Clang 11.0.0 (clang-1100.0.33.8)]

**Comment by rossbar** [comment]
I'm not very familiar with `enum` so I don't really understand the `enum.EnumMeta` type that you get from `type(StrEnum)`. However - one thing that stands out immediately is that `StrEnum` inherits from `str` which is the root of the behavior in this case. If you define an Enum that doesn't inherit from `str`, you get the behavior you expect:

```python
>>> class DifferentEnum(Enum):
        A = 'a'
        B = 'bb'
>>> np.array([DifferentEnum.A])
array([<DifferentEnum.A: 'a'>], dtype=object)
```

**Comment by seberg** [comment]
We currently also coerce subclasses of types that NumPy knows about to store them into arrays. I agree with it being bad behaviour, in the long run, I would like to deprecate it.

**Comment by eric-wieser** [comment]
I don't think we can deprecate it without providing and documenting some way for users to opt-in explicitly. Perhaps requiring they implement `__array__` is reasonable (as returning a 0d str array in this case).

**Comment by seberg** [comment]
Yeah, implementinug `__array__` would be an easy workaround. Note that we would only deprecate the automatic discovery, `dtype=np.str_` would still work of course, but I agree that it might have a bit more impact (at least for the basic python types).

**Comment by alonme** [comment]
@seberg Do you have an idea for a solution? will it be simple to implement? can i help?

**Comment by seberg** [comment]
Well, if you are in for the long haul, I would be very happy to ping you for review eventually :).

If you are not. You *can* look at the dtype discovery code and try to add the deprecation right now It should be fairly straight forward (if you are familiar with the Python C-API and not scared of getting your hands dirty), since you just need to add `Py_TYPE(...) == ...` checks and DeprecationWarnings if it is true. I hope to restructure that code sooner rather then later, but that does not mean we cannot add the deprecation warning right now (it just means I am not super motivated to do it myself).
We may have to sign such a change of on the mailing list, and since I do not know if someone may rely on such behaviour, there is a chance that we will not be able to do it, or have to think about how to make a transition more gentle.

**Comment by alonme** [comment]
Not very familiar with the C-API but will be happy to get my hands dirty.
Ill try and take a look.

**Comment by Noxaster** [comment]
Hi @seberg and everyone, I've confirmed this is still an issue on the latest main (2.5.0.dev).

Actually, it's a bit worse than just a `dtype` mismatch. As the author mentioned:
> ... the data saved is X chars from the string representation of the enum.

Here is an example:
```python
>>> import numpy as np
>>> from enum import Enum
>>> class StrEnum(str, Enum):
...     A = 'aa'
...
>>> arr = np.array([StrEnum.A])
>>> print(arr.dtype)
<U2
>>> print(arr)
['St']
```
I’ve tracked the discovery logic to `discover_dtype_from_pyobject` in `array_coercion.c`. Interestingly, I noticed there's already a disabled `if (0)` block there with a `TODO` for handling scalar subclass warnings.

I'm planning to enable and refine this logic to address the issue. I will also add regression tests with `xfail` cases to document the current data corruption behavior. Does this approach look good to you? I can open a PR once you confirm.

**Comment by seberg** [comment]
The point is that this subclass of string doesn't behave like a string.  You can fix that but how do you do it without adding a layer of special casing, which seems very dubious to me?
Sure, we could go to object dtype here, but that is a behavior change as well for subclasses where things work.

**Comment by Noxaster** [comment]
Thanks for the feedback! I clearly didn't explain myself well. I'm definitely not suggesting any special casing or behavior changes.

I simply want to flip the switch on the generic `FutureWarning` logic that is already present in the `array_coercion.c` (currently disabled):
```c
/* TODO: Enable warning about subclass handling */
        if ((0) && !((*flags) & GAVE_SUBCLASS_WARNING)) {
            if (DEPRECATE_FUTUREWARNING(
                    "in the future NumPy will not automatically find the "
                    "dtype for subclasses of scalars known to NumPy (i.e. "
                    "python types). Use the appropriate `dtype=...` to create "
                    "this array. This will use the `object` dtype or raise "
                    "an error in the future.") < 0) {
                return NULL;
            }
            *flags |= GAVE_SUBCLASS_WARNING;
        }
```
Also the warning logic is safely located at the end of the discovery chain.

My plan is just to remove that `(0) &&`. This activates the planned warning for all scalar subclasses (e.g. `IntEnum`), not just `StrEnum`.

If this sounds reasonable, I can open a PR to enable this and add `xfail` regression tests to document the current behavior for future possible fixes.
